File: Surrounded Regions.cpp
#include <iostream>
#include <vector>

class Solution {
  public:
    void solve(std::vector<std::vector<char>> &board) {
        int m = board.size();
        int n = board[0].size();

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                bool ok = canReachBorder(board, i, j);
                if (ok) {
                    set(board, i, j, 'Y');
                } else {
                    set(board, i, j, 'N');
                }
            }
        }

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (board[i][j] == 'Y') {
                    board[i][j] = 'O';
                }
                if (board[i][j] == 'N') {
                    board[i][j] = 'X';
                }
            }
        }
    }

    bool canReachBorder(std::vector<std::vector<char>> &board, int i, int j) {
        if (board[i][j] != 'O') {
            return false;
        }
        board[i][j] = 'V';

        if ((i == 0 || i == board.size() - 1 || j == 0 ||
             j == board[0].size() - 1)) {
            return true;
        }

        return (canReachBorder(board, i + 1, j) ||
                canReachBorder(board, i - 1, j) ||
                canReachBorder(board, i, j + 1) ||
                canReachBorder(board, i, j - 1));
    }

    void set(std::vector<std::vector<char>> &board, int i, int j, char c) {
        if ((i < 0 || i >= board.size() || j < 0 || j >= board[0].size()) ||
            board[i][j] != 'V') {
            return;
        }
        board[i][j] = c;

        set(board, i + 1, j, c);
        set(board, i - 1, j, c);
        set(board, i, j + 1, c);
        set(board, i, j - 1, c);
    }
};

File: Surrounded Regions.java
class Solution {
    public void solve(char[][] board) {
        int m = board.length;
        int n = board[0].length;

        for (int j = 0; j < n; ++j) {
            dfs(board, 0, j, m, n);
            dfs(board, m - 1, j, m, n);
        }
        for (int i = 0; i < m; ++i) {
            dfs(board, i, 0, m, n);
            dfs(board, i, n - 1, m, n);
        }

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (board[i][j] == 'O') {
                    board[i][j] = 'X';
                }
            }
        }
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (board[i][j] == 'S') {
                    board[i][j] = 'O';
                }
            }
        }
    }

    private void dfs(char[][] board, int i, int j, int m, int n) {
        if (i < 0 || i >= m || j < 0 || j >= n || board[i][j] != 'O') {
            return;
        }
        board[i][j] = 'S';

        dfs(board, i + 1, j, m, n);
        dfs(board, i - 1, j, m, n);
        dfs(board, i, j + 1, m, n);
        dfs(board, i, j - 1, m, n);
    }
}
